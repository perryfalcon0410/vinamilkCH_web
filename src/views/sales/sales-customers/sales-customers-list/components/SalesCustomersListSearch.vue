<template>

  <!-- START - Search -->
  <b-form
    @keydown.enter="onClickSearchButton"
  >
    <v-card-actions
      title="Tìm kiếm"
      class="d-print-none"
    >

      <!-- START - Phone -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="mt-sm-1 mt-xl-0"
        >
          Di động
        </div>
        <b-input-group
          class="input-group-merge"
        >
          <b-form-input
            v-model.trim="phoneNumber"
            data-test="phoneNumber"
            autocomplete="on"
            placeholder="Nhập chính xác 4 số cuối"
            maxlength="10"
            @keypress="$onlyNumberInput"
          />
          <b-input-group-append
            is-text
          >
            <b-icon-x
              v-show="phoneNumber"
              class="cursor-pointer text-gray"
              @click="phoneNumber = null"
            />
          </b-input-group-append>
        </b-input-group>
      </b-col>
      <!-- END - Phone -->

      <!-- START - Full Name -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="mt-sm-1 mt-xl-0"
        >
          Khách hàng
        </div>
        <b-input-group
          class="input-group-merge"
        >
          <b-form-input
            v-model="customerName"
            data-test="customerName"
            placeholder="Nhập họ tên/mã"
          />
          <b-input-group-append
            is-text
          >
            <b-icon-x
              v-show="customerName"
              class="cursor-pointer text-gray"
              @click="customerName = null"
            />
          </b-input-group-append>
        </b-input-group>
      </b-col>
      <!-- END - Full Name -->

      <!-- START - Group -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="mt-sm-1 mt-xl-0"
        >
          Nhóm khách hàng
        </div>
        <tree-select
          v-model="customerTypesSelected"
          data-test="customerTypes"
          :options="customerTypeOptions"
          :auto-load-root-options="false"
          :searchable="false"
          placeholder="Tất cả"
          no-options-text="Không có dữ liệu"
        />
      </b-col>
      <!-- END - Group -->

      <!-- START - Status -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="mt-sm-1 mt-xl-0"
        >
          Trạng thái
        </div>
        <tree-select
          id="status"
          v-model="statusSelected"
          :options="statuOptions"
          :searchable="false"
          placeholder="Tất cả"
          no-options-text="Không có dữ liệu"
        />
      </b-col>
      <!-- END - Status -->

      <!-- START - Gender -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="mt-sm-1 mt-xl-0"
        >
          Giới tính
        </div>
        <tree-select
          id="genders"
          v-model="gendersSelected"
          :options="genderOptions"
          :searchable="false"
          placeholder="Tất cả"
          no-options-text="Không có dữ liệu"
        />
      </b-col>
      <!-- END - Gender -->

      <!-- START - Location -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="mt-sm-1 mt-xl-0"
        >
          Khu vực
        </div>
        <tree-select
          id="areas"
          v-model="areasSelected"
          :options="areaOptions"
          :auto-load-root-options="false"
          placeholder="Tất cả"
          no-options-text="Không có dữ liệu"
          no-results-text="Không tìm thấy kết quả"
        />
      </b-col>
      <!-- END - Location -->

      <!-- START - Private customer -->
      <b-col
        xl
        lg="3"
        sm="4"
        class="h7"
      >
        <div
          class="text-white mt-sm-1 mt-xl-0"
          onmousedown="return false;"
          style="cursor: context-menu; margin-bottom: 5px;"
        >
          Khách hàng riêng
        </div>
        <b-form-checkbox
          v-model="privateCustomer"
          class="text-nowrap"
        >
          Khách hàng của cửa hàng
        </b-form-checkbox>

      </b-col>
      <!-- END - Private customer -->

      <!-- START - Search button -->
      <b-col
        xl
        lg="3"
        sm="4"
      >
        <!--"onmousedown" is prevent hightlight text -->
        <div
          class="text-white h7"
          onmousedown="return false;"
          style="cursor: context-menu;"
        >
          Tìm kiếm
        </div>
        <b-button
          id="btnSearch"
          class="btn-brand-1 h8 align-items-button-center mt-sm-1 mt-lg-0"
          variant="someThing"
          @click="onClickSearchButton()"
        >
          <b-icon-search class="mr-50" />
          Tìm kiếm
        </b-button>
      </b-col>
      <!-- END - Search button -->
    </v-card-actions>
  </b-form>
  <!-- END - Search -->
</template>

<script>
import VCardActions from '@core/components/v-card-actions/VCardActions.vue'
import {
  mapActions,
  mapGetters,
} from 'vuex'
import commonData from '@/@db/common'
import customerData from '@/@db/customer'

import {
  CUSTOMER,
  // GETTERS
  SHOP_LOCATIONS_SEARCH_GETTER,
  CUSTOMER_TYPES_GETTER,
  // ACTIONS
  GET_SHOP_LOCATIONS_SEARCH_ACTION,
} from '../../store-module/type'

export default {
  components: {
    VCardActions,
  },

  data() {
    return {
      customerName: null,
      phoneNumber: null,
      customerTypesSelected: null,
      statuOptions: customerData.status,
      statusSelected: customerData.status[0].id,
      genderOptions: commonData.genders,
      gendersSelected: null,
      areasSelected: null,
      privateCustomer: true,
      customerTypeOptions: null,
      apiStatus: { // success or fail
        // customerTypes: false,
        shopLocationsSearch: false,
      },

      // decentralization
      decentralization: {
        formId: 1,
        ctrlId: 1,
      },
    }
  },

  computed: {
    ...mapGetters(CUSTOMER, [
      SHOP_LOCATIONS_SEARCH_GETTER,
      CUSTOMER_TYPES_GETTER,
    ]),
    getCustomerTypeOptions() {
      if (this.CUSTOMER_TYPES_GETTER) {
        return this.CUSTOMER_TYPES_GETTER.map(data => ({
          id: data.id,
          label: data.name,
        }))
      }
      return []
    },
    areaOptions() {
      if (this.SHOP_LOCATIONS_SEARCH_GETTER) {
        return this.SHOP_LOCATIONS_SEARCH_GETTER.map(data => ({
          id: data.id,
          label: data.provinceAndDistrictName,
          default: data.default,
        }))
      }
      return []
    },
    // apiStatusResult() {
    //   return Object.values(this.apiStatus).every((val, i, arr) => val && arr[0])
    // },
  },

  watch: {
    areaOptions() {
      this.areaSelectedDefault()
    },
    // apiStatusResult() {
    //   if (this.apiStatusResult) {
    //      this.onSearch()
    //   }
    // },
    getCustomerTypeOptions() {
      this.customerTypeOptions = [...this.getCustomerTypeOptions]
    },
  },

  mounted() {
    this.GET_SHOP_LOCATIONS_SEARCH_ACTION({
      data: { ...this.decentralization },
      onSuccess: () => {
        // this.apiStatus.shopLocationsSearch = true
      },
    })
  },

  methods: {
    ...mapActions(CUSTOMER, [
      GET_SHOP_LOCATIONS_SEARCH_ACTION,
    ]),
    areaSelectedDefault() {
      if (this.SHOP_LOCATIONS_SEARCH_GETTER) {
        this.areasSelected = this.SHOP_LOCATIONS_SEARCH_GETTER.find(e => e.default === true).id
        this.onSearch()
      }
    },
    onSearch() {
      const searchData = {
        searchKeywords: this.customerName?.trim(),
        phoneNumber: this.phoneNumber,
        customerTypeId: this.customerTypesSelected,
        status: this.statusSelected,
        genderId: this.gendersSelected,
        areaId: this.areasSelected,
        isShop: this.privateCustomer,
      }
      this.getCustomers(searchData)
    },
    onClickSearchButton() {
      this.onSearch()
    },
    getCustomers(data) {
      this.$emit('onSearchClick', data)
    },
  },
}
</script>
