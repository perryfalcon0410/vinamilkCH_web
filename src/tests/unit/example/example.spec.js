import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import BootstrapVue from 'bootstrap-vue'
import SalesCustomersListSearch from '@/views/sales/sales-customers/sales-customers-list/components/SalesCustomersListSearch.vue'
import TreeSelect from '@riophae/vue-treeselect'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(BootstrapVue)
localVue.use(TreeSelect)

describe('SalesCustomersListSearch', () => {
  let actions
  let getters
  let store
  let wrapper
  beforeEach(() => {
    actions = {
      onClickSearchButton: jest.fn(),
    }
    getters = {
      SHOP_LOCATIONS_SEARCH_GETTER: () => [
        {
          areaCode: '760',
          areaName: 'Quận 1',
          createdAt: '2021-03-30T10:31:59.428',
          createdBy: null,
          default: false,
          id: 49,
          provinceAndDistrictName: 'TP Hồ Chí Minh - Quận 1',
          updatedAt: null,
          updatedBy: null,
        },
        {
          areaCode: '001',
          areaName: 'Quận Ba Đình',
          createdAt: '2021-03-30T10:31:59.428',
          createdBy: null,
          default: false,
          id: 47,
          provinceAndDistrictName: 'Hà Nội - Quận Ba Đình',
          updatedAt: null,
          updatedBy: null,
        },
      ],
    }
    store = new Vuex.Store({
      actions,
      getters,
    })
    wrapper = shallowMount(SalesCustomersListSearch, { store, localVue })
  })

  it('test field phoneNumber of SalesCustomerListSearch', async () => {
    await wrapper.setData({ phoneNumber: '0364121908' })
    // phoneNumber
    const fieldPhoneNumber = wrapper.find('[data-test="phoneNumber"]')
    expect(fieldPhoneNumber.exists()).toBe(true)
    expect(fieldPhoneNumber.element.attributes.getNamedItem('value').value).toEqual('0364121908')
  })
  it('test field customerName of SalesCustomerListSearch', async () => {
    await wrapper.setData({ customerName: 'Nguyen Van A' })
    const fieldCustomerName = wrapper.find('[data-test="customerName"]')
    expect(fieldCustomerName.exists()).toBe(true)
    expect(fieldCustomerName.element.attributes.getNamedItem('value').value).toEqual('Nguyen Van A')
  })

  it('has a button search on SalesCustomerListSearch', async () => {
    expect(wrapper.contains('#btnSearch')).toBe(true)
    const btnSearch = wrapper.find('#btnSearch')
    const onClickSearchButton = jest.fn()
    wrapper.setMethods({ onClickSearchButton })
    // const spy = jest.spyOn(wrapper.vm, 'onClickSearchButton')
    // btnSearch.trigger('click')
    // await wrapper.vm.$nextTick()
    // expect(spy).toHaveBeenCalled()
    // jest.restoreAllMocks()
    btnSearch.trigger('click')
    expect(onClickSearchButton).toHaveBeenCalledTimes(1)
  })

  it('has class on search button', () => {
    const btnSearch = wrapper.find('#btnSearch')
    expect(btnSearch.contains('.btn-brand-1')).toBe(true)
    expect(btnSearch.contains('.h8')).toBe(true)
    expect(btnSearch.contains('.align-items-button-center')).toBe(true)
  })

  it('test snapshoot of SalesCustomerListSearch.vue', () => {
    expect(wrapper).toMatchSnapshot()
  })

  // it('test field customerTypes of SalesCustomerListSearch', async () => {
  //   await wrapper.setData({
  //     customerTypeOptions: [
  //       { id: '0', label: 'Khách hàng thân thiên DLĐộng' },
  //       { id: '1', label: 'Khách hàng thường' },
  //       { id: '2', label: 'Khách hàng VIP' },
  //     ],
  //     customerTypesSelected: '1',
  //   })
  //   const fieldCustomerTypes = wrapper.find('[data-test="customerTypes"]')
  //   expect(fieldCustomerTypes.exists()).toBe(true)
  //   expect(fieldCustomerTypes.forest).toEqual('Khách hàng thường')
  //   expect(fieldCustomerTypes.vm).toEqual(['anything'])
  // })
})
