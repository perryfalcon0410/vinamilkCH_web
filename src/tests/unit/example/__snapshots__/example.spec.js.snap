// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesCustomersListSearch test snapshoot of SalesCustomerListSearch.vue 1`] = `
<b-form-stub>
  <v-card-actions-stub title="Tìm kiếm" cancollapse="true" expanddefault="true" class="d-print-none">
    <b-col-stub lg="3" sm="4" tag="div" xl="true" class="h7">
      <div class="mt-sm-1 mt-xl-0">
        Di động
      </div>
      <b-input-group-stub tag="div" class="input-group-merge">
        <b-form-input-stub autocomplete="on" debounce="0" formatter="[Function]" placeholder="Nhập chính xác 4 số cuối" type="text" data-test="phoneNumber" maxlength="10"></b-form-input-stub>
        <b-input-group-append-stub istext="true" tag="div">
          <b-icon-x class="cursor-pointer text-gray" style="display: none;"></b-icon-x>
        </b-input-group-append-stub>
      </b-input-group-stub>
    </b-col-stub>
    <b-col-stub lg="3" sm="4" tag="div" xl="true" class="h7">
      <div class="mt-sm-1 mt-xl-0">
        Khách hàng
      </div>
      <b-input-group-stub tag="div" class="input-group-merge">
        <b-form-input-stub debounce="0" formatter="[Function]" placeholder="Nhập họ tên/mã" type="text" data-test="customerName"></b-form-input-stub>
        <b-input-group-append-stub istext="true" tag="div">
          <b-icon-x class="cursor-pointer text-gray" style="display: none;"></b-icon-x>
        </b-input-group-append-stub>
      </b-input-group-stub>
    </b-col-stub>
    <b-col-stub lg="3" sm="4" tag="div" xl="true" class="h7">
      <div class="mt-sm-1 mt-xl-0">
        Nhóm khách hàng
      </div>
      <tree-select data-test="customerTypes" placeholder="Tất cả" no-options-text="Không có dữ liệu"></tree-select>
    </b-col-stub>
    <b-col-stub lg="3" sm="4" tag="div" xl="true" class="h7">
      <div class="mt-sm-1 mt-xl-0">
        Trạng thái
      </div>
      <tree-select id="status" options="[object Object],[object Object]" placeholder="Tất cả" no-options-text="Không có dữ liệu"></tree-select>
    </b-col-stub>
    <b-col-stub lg="3" sm="4" tag="div" xl="true" class="h7">
      <div class="mt-sm-1 mt-xl-0">
        Giới tính
      </div>
      <tree-select id="genders" options="[object Object],[object Object],[object Object]" placeholder="Tất cả" no-options-text="Không có dữ liệu"></tree-select>
    </b-col-stub>
    <b-col-stub lg="3" sm="4" tag="div" xl="true" class="h7">
      <div class="mt-sm-1 mt-xl-0">
        Khu vực
      </div>
      <tree-select id="areas" options="" placeholder="Tất cả" no-options-text="Không có dữ liệu" no-results-text="Không tìm thấy kết quả"></tree-select>
    </b-col-stub>
    <div class="h7 mx-50">
      <div onmousedown="return false;" class="text-white mt-sm-1 mt-xl-0 mb-45" style="cursor: context-menu;">
        Khách hàng riêng
      </div>
      <b-form-checkbox-stub checked="true" value="true" class="text-nowrap">
        Khách hàng của cửa hàng
      </b-form-checkbox-stub>
    </div>
    <div class="mt-auto mx-35">
      <b-button-stub tag="button" target="_self" type="button" variant="someThing" id="btnSearch" class="btn-brand-1 h8 align-items-button-center mt-sm-1 mt-lg-0">
        <b-icon-search class="mr-50"></b-icon-search>
        Tìm kiếm
      </b-button-stub>
    </div>
  </v-card-actions-stub>
</b-form-stub>
`;
